rholang/src/main/k/rholang.k

```diff
@@ -0,0 +1,132 @@
+module RHO-SYNTAX
+imports SUBSTITUTION
+
+// Syntax for processes
+syntax Proc ::=
+              // Empty process
+                "Nil"
+              // Listen
+              | "for(" Name "<-" Name "){" Proc "}" [binder]
+              // Persistent listen
+              | "for(" Name "<=" Name "){" Proc "}" [binder]
+              | "contract" Name "(" Name ")" "=" "{" Proc "}" [binder]
+              // Send
+              | Name "!(" Proc ")"
+              // Evaluate
+              | "*" Name
+              // New
+              | "new" Id "in" "{" Proc "}" [binder]
+              // Parallel
+              | Proc "|" Proc [left]
+              // Misc
+              | Exp
+
+// Syntax for Names; we allow variables to take the place of names, but note
+// that no rholang program will compile if there are free variables left in
+// the program. Thus the variable must be bound to a name of the form @Proc
+syntax Name ::= "@" Proc
+              | Id
+
+
+// We have to add Higher Processes (HigherProc) as well as Chan (Channels— this
+// word choice is perhaps unfortunate, as in rholang channels and names are the
+// same). Thus far in the K-framework we will allow channels to be names, but also
+// the channels which spawn from the "new" construct. The higher processes are
+// processes which allow channels to be substituted into the places that names
+// used to be--i.e. the syntax supports the "new" channel construct.
+//
+// You'll note that in the configuration below, only programs of the syntactic
+// category Proc can be written; this means that one cannot write the syntax that
+// K gives for new channels, even though the syntax is considered correct. This
+// is to make the "new" channels unforgeable.
+syntax HigherProc ::=
+                "for(" Chan "<-" Chan "){" HigherProc "}" [binder]
+              | "for(" Chan "<=" Chan "){" HigherProc "}" [binder]
+              | "contract" Chan "(" Chan ")" "=" "{" HigherProc "}" [binder]
+              | Chan "!(" HigherProc ")"
+              | "*" Chan
+              | "new" Id "in" "{" HigherProc "}" [binder]
+              | Proc
+              | "#(" Int ")"
+              | HigherProc "|" HigherProc [left] // should this be changed to a different category? Procs?
+              | "{" HigherProc "}" [bracket]
+
+syntax Chan ::= "@" HigherProc
+              | Name
+
+syntax Exp  ::= AExp
```

Comment:
`Exp` is structurally equivalent to the Nil process, right? Should we roll these expressions and the `Nil` into one syntax grouping?

Reply:
Commenting is not distracting, please keep doing so—all comments/collaborations are very welcome. It's always useful to have an extra set of eyes looking at this.

That Exp and Nil are structurally equivalent is an important insight; I hadn't thought about changing the syntax in that way, but that's a good idea. Doing so simplifies the formal semantics.

Reply:
Also, with regards to using `substitution`, I don't need to do anything to `kompile` to get it to run, but now that I'm looking at it, it looks like I might be using an older version of K.

Reply:
I'm having troubles with k5 built from the latest commit. Might just be me though.. What version are you using / where did you get it?

Reply:
Also, I don't think Exp and Nil are structurally equivalent, maybe semantically equivalent? Structurally equivalent processes when quoted should create equivalent names. I don't believe we should expect a `for(x <- @"hello")` to match with a `@"goodbye"!(P)`.

Reply:
Exp and Nil are neither structurally equivalent nor semantically equivalent.  It's just the case that like `new x in { x!("Hi there") }`, it can have no effect when run in parallel with another process.  Exp and Nil are only semantically equivalent *in a par context*.

Reply:
I'm using the K VM from kframework.org. Now that I'm looking at it, I need to get a more recent version; I'm on 3.4. 😑

Reply:
@differentialderek Let me know if you get the nightly build of k5 working? I am getting gcc errors

Reply:
Got k5 working, here is how I installed it: https://github.com/Jake-Gillberg/kframework-env/blob/master/Dockerfile#L36